### ISSUE TITLE
Need to track purchases in google analytics

### ISSUE THREAD

**Issue by orangedrink** [Tizra/Tizra-Customer-Tracker#71]
I need to track purchases in Google analytics and would like to set this up as a goal with the url being the receipt or confirmation page and would also like to have funnel steps for the paywall and each step of the commerce transaction but would need a unique url for each of those pages or at least a different layout (so that I can change out the url being pushed into analytics) 

Is this possible or is there a better way to measure purchases within analytics?

**Comment by daviddurand** [comment]
There are a set of special Google Analytics "goals" (virtual pages), that correspond to different stages of the commerce process, and various other events of interest. Configuring the use of these is a bit annoying, and the way google tracks events is by having special "event URLs" that we send tracking information for when pages are displayed, or links are clicked.

Here are most of them (I'm not in analytics all the time, but last I talked with Abe, the settings mentioned here were the ones to use for some of these URLs):

```
Match Type: Head Match
Goal URL: /commerce/success
Goal Name: Successfully Complete Purchase
Case Sensitive: unchecked
Goal Value: 0.0
```

```
Step 1
URL: /commerce/form/paymentmethod
usually paypal or paypalpro
Name: 1-Arrive at Payment Form
(required is checked)
```

```
Step 2
URL: /commerce/verify/paymentmethod
Name: 2-Arrive at Confirmation Page
```

```
Step 3:
URL: /commerce/success/packageURLId
(there may be more than one of these if the cart contains multiple products).
```

Here are some of the other tracking tags:

```
/offer/link/offerUrlId
when a person clicks on an offer link to see the full offer description page
```

```
/commerce/cart/login
when a person clicks the login link displayed by the shopping cart
```

```
/commerce/cart/register
when a person clicks the registration link displayed by the shopping cart.
```

```
/commerce/add_cart/offerUrlId
when a person adds an offer to a cart.
```

```
/commerce/fail/paymentmethod
when a transaction fails due to an error (usually a bad credit card).
```

```
/registeruser/success/emailsent
User registered successfully, confirmation email was sent.
```

```
/registeruser/activate/fail/UrlIdofContentPage
User failed to activate account from an email activation page
```

```
/offer/acceptfree/offerUrlId
User accepted a free offer by clicking on accept link.
```

```
/offer/lock/display/offerUrlId
The form to unlock a locked offer was displayed
```

```
/free/success/offerUrlId
A free offer was accepted
```

```
/free/failed/offerUrlId
A free offer could not be accepted (because had already been used).
```

```
/cart/display/offerUrlId
The cart displayed an offer. May occur multiple times in one cart display.
```

```
/cart/offer/remove/offerUrlId
An offer was removed from the cart.
```

**Comment by orangedrink** [comment]
That's exactly what I was looking for. 

Unfortunately I'm not seeing these URLs in GA because I had to write some custom code on the TRB site to do some virtual URLs of my own on other pages. To accomplish that I had to remove our tracking code from the reporting tab and add my own custom tracking code in a free html block.

I'll see if I can figure out how you guys did it in your code and emulate that. If not I'll shout for some more help.

Thanks David

**Comment by daviddurand** [comment]
I'd recommend that whichever tracking code you want to use for tracking purchases should be the one that you enter into the Tizra system, and that you not try to reimplement it. The recognition of what pages are being delivered is built into the template logic of the system on the backend, and not in the scripting layer and at least some of those tracking functions are implemented by rewriting links as they are generated by the templates. Trying to do it after the fact in Javascript will be a real bear, both in the sense that it will be hard to do, and that it may be missing some degree of functionality and reliability.

In case this is an issue, you should know that we don't get any particular access the accounts that you configure into the system -- all the GA tracking that we do for system monitoring, etc. is done through a separate account that spans all Tizra sites.

If you want to add your own virtual URLs, you may be able just use GA Javascript that we include onto the pages -- it's totally standard in any case. I wish I could give you some of our utility functions, but they are implemented in the templates, and not in javascript.

**Comment by orangedrink** [comment]
Well it seems like my hands are tied on this one since the page view is tracked before any of my code comes into play. 

Would it be possible to keep this code in the header:

```
    _gaq.push(
        ['_setAccount', 'UA-xxxxx-1'], 
        ['_setDomainName', 'none'],
        ['_setAllowLinker', true]
    )} catch(err) {}
```

But then in the end of the page do something like this:

```
    if( typof(ga_url) != "undefined" ){
        try {_gaq.push(['_trackPageview'], ga_url);} catch(err) {}
    }else{
        try {_gaq.push(['_trackPageview']);} catch(err) {}
    }
```

This way if I create a variable called ga_url and assign it a value anywhere in the layout or in my scripts then that is the url that gets pushed into GA. If I don't create this variable everything works as it does currently.

Does that make sense? Any other ideas how to solve this one?

**Comment by orangedrink** [comment]
TRB says they will renew for another year if we can make this happen ;) any hope of making this work by 2013?

**Comment by daviddurand** [comment]
I'm sorry, did not mean to drop this...

you should just your account into the Tizra system, and call the javascript we supply with your custom events. Our scripts are just the standard Google stuff anyway. You can call `_gaq.push`, and all the rest of the google scripts yourself if you want to for your custom events. If there are loading order problems, you can use some event handlers to start them at the right time.

The sales tracking is moderately complicated, and the server knows how to generate the right stuff to track all the events, and the ways that users could invoke them -- important for tuning the interface. The server will also continue to generate this kind of information even if we make any changes in the commerce flow -- something that you won't be able to do if you re-build the wheel on this one.

Basically you should be able to get custom events without the customizations that are blocking the e-commerce tracking we built. You could implement scripts to duplicate what we did yourself, but it would just be a waste of effort as far as I can see.

Call me if I'm missing something.

**Comment by orangedrink** [comment]
Thanks David. The reason that I rebuilt this was so that I could create custom URLs. Events weren't working so well for us. 

I could definitely put the tracking code back into Tizra and call `_gaq.push` but that would result in the page being tracked twice.

**Comment by daviddurand** [comment]
We're using custom URLs as well -- there's really not much difference between custom URLs and events, really. Regardless of what additional tracking you want to add, events or URLs, you don't need to build your own separate tracking, as we are both calling Google. `_gaq` isn't mine, it's just the (faster and now recommended) _asynchronous_ Analytics API.

[Google's asynchronous event API for analytics](https://developers.google.com/analytics/devguides/collection/gajs/methods/gaJSApi_gaq)

I'm not suggesting that you add the original tizra tracking to your customization (which _would_ track things double), but that you remove your code entirely, configure Tizra to talk the the customer's analytics account, and make your google analytics calls using the google scripts we have already loaded. We already call two separate GA accounts for each event (the Tizra system one, and any customer one that is configured), so you can look at the Tizra page source for examples of the scripts you may want to write.

The current duplication adds unnecessary network traffic and script rending delay, will be a an ongoing maintenance burden -- and because it ignores the Tizra tracking info, means that you can't get the custom URLs and events that we generate for the e-commerce links.

For instance, one of the things that we track, that you can't, is form submissions, where we have modified some scripts to preserve google tracking info. Since we generate those scripts on the fly, you'd have to do some complicated markup surgery.

**Comment by orangedrink** [comment]
"remove your code entirely, configure Tizra to talk the the customer's analytics account, and make your google analytics calls using the google scripts we have already loaded."

Wouldn't this still result in `_trackPageview` being called twice? once by your code and once with my custom URL?

The only real option I see here is the code snippet I posted above. Not sure if you saw that or not...

**Comment by daviddurand** [comment]
When you track a custom URL, The custom URL is the only URL that gets counted -- it doesn't matter what page it's called from. If you want to track multiple things on a single page, you just keep calling the tracker with different URLs...

As long as you don't track the page's URL again yourself, it won't be tracked more than once. If you look at the HTML for any page that contains offers in an offer block, you will see that first the page is tracked, and then each offer gets tracked (to record that it was displayed to the user). The subsequent series of calls on the special offer URLs don't increase the count of the page itself.

I couldn't understand the point of the script, though I did see it. If your point was to only call `_trackPageView` once, I can see why something like that would seem useful.

But you don't need to ensure that you only call it once. In fact for each distinct thing that you might want to track in a pipeline, for instance, you will need a call. We actually note both _events_ and _custom URLs_  for some things, given the occasional annoyances of doing fancy stuff with GA.

You do have to make sure that any URLs for tracking things like offers, for instance, can be distinguished from actual page views so that you can get aggregate visit statistics that are sensible.

**Comment by orangedrink** [comment]
Thanks David.

"If you want to track multiple things on a single page, you just keep calling the tracker with different URLs..."
Calling _trackPageview more than once per page will inflate unique pageview statistics.

"As long as you don't track the page's URL again yourself, it won't be tracked more than once."
I need to construct a custom URL in my script and push it to GA. This is the whole point. The code I suggested above would allow me to customize the URL before you push it to GA. 

Your offers are generating events not pageviews. the method you use is also pretty unreliable since it results in a race condition between the event being pushed into GA and the next page loading.

"But you don't need to ensure that you only call it once. In fact for each distinct thing that you might want to track in a pipeline, for instance, you will need a call. We actually note both events and custom URLs for some things, given the occasional annoyances of doing fancy stuff with GA."
We aren't using events. We only want to use custom URLs and we want our pageview statistics to be reliable.

**Comment by daviddurand** [comment]
I've told you how to push a custom URL to GA, if you want to. You still haven't explained why you want to do this; if you could give me a reason it might help me to understand the tradeoffs. It looks like you might be just translating the URL; if so, GA has a feature for automatically rewriting URLs before doing statistics or anything else on them. It's poorly named **Filters**, but it's a URL rewriting and normalization step, _not_ a filter query. I provide some pointers below. I think your custom URLs could be replaced by this feature very easily.

Is this your custom tracking code?

```
var ga_url=window.location.href.replace(/^https?:\/\/(.+?)\/(.*)/gi,'$2');
var _gaq = _gaq || [];
_gaq.push(['_setAccount', 'UA-20344518-1']);
```

All it seems to do is to delete the protocol and the server name from the URL. We track the full URL because the information is useful for filtering out the staging site, etc.

We use GA "filters" (in conjunction with complete URLs) to normalize incoming URLs to create different kinds of reports from the same tracking information. You can also configure multiple dashboards on the same underlying tracking ID. Here's how to use filters:

[Google Analytics Filters](http://support.google.com/googleanalytics/bin/answer.py?hl=en&answer=55494)

(follow the "tracking multiple subdomains" link, if you want to see some of the ways we use this capability).

You can also filter the page view statistics themselves (with a _real_ filtering operation), by restricting the URLs that contribute to some reports. We have used the URL "filtering" to put site information in the url paths to get both aggregated and individual site statistics easily. This is another way to ignore "duplicates", if you want to. 

I'm not sure what you mean by a race condition, as the browser is single threaded. I certainly don't appreciate digs about  "unreliable" tracking, when I went to significant trouble to rewrite all our tracking code to use the tracking method that Google recommends in their current documentation. That was one reason I linked to the doc, because at first I found the style rather weird, especially the part where `_gaq` starts as an array and turns into an object once the asynchronous javascript fetch finishes.

Your analysis of what we send to GA, is, by the way, not correct. I just checked the source, and, here are some of the cases we take care of. `_trackPageView` is called for the following situations because events cannot be used as "goals" in the sales funnel tracking functions:

```
expanded view / mobile view /printable view etc.
registeruser
commercefailed/success/freesuccess/verify,  etc.
cart/display
```

I use events for things like:

```
cart
attempt to accept free offer (the link).
registeruser (yes, I log this both ways!).
```

This was the result of weeks of (part-time) reading and experimentation in collaboration with Abe who is our big GA user, and then several days to a week of implementation, integration and testing, since it affects 20-30% of all page templates, and only the more complicated among them. In other words it's something that I'm not going to change lightly.

**Comment by orangedrink** [comment]
That wasn't meant to be a dig. Pushing events on a the click of an &lt;a> tag only works if the ajax operation competes before the page pointed to by the href attribute is loaded. You may also push these as urls elsewhere but in the onclick handler of these links you are using events. This is the code I was referring to.
`<a href="http://amonline.trb.org/1tcldk/?offerControlledObjectId=64375990" 
onclick="try {_gaq.push(['_setAccount', 'UA-1587864-1'], ['_trackEvent', 'offer/link/', '1tcldk', 'value', 500 ] );} catch(err) {}" class="t-offer-link">
Offer to the general public on All 2012 and 2011 Documents </a>`

I understand that you put a lot of time and effort into developing your Google Analytics code but there's always a way to improve on things. When you have suggestions for me I am glad to consider them. Please don't take it as a personal attack if I point out ways that I feel some of your code could be improved. I'm sure you would agree that Tizra isn't perfect and without feedback there are a lot of ways you will never be able to make it better.

Aaanyway, what I need to do is construct the rewritten GA URLs from various pieces of metadata. For document views I'm constructing a url that looks something like this in the end "/2012/Paper/648/12-0076". All of the parts between the slashes are metadata fields. This is handy for the content drilldown area of Google Analytics and gives TRB all the information they need in whichever dashboard widget or report they choose to use. 

I know that each document has a url property that we could populate but this wouldn't work as desired in the content drilldown area of analytics since slashes don't seem to be allowed. Also there is no good way to retroactively populate so many of these existing documents with urls although if I'm not mistaken you do owe us a content load or two.
